tools/src/bin/build-blame.rs

```diff
@@ -0,0 +1,357 @@
+    }
+    let mut linecount = 0;
+    for b in data {
+        if *b == b'\n' {
+            linecount += 1;
+        }
+    }
+    if data[data.len() - 1] != b'\n' {
+        linecount += 1;
+    }
+    linecount
+}
+
+fn unmodified_lines(
+    blob: &git2::Blob,
+    parent_blob: &git2::Blob,
+) -> Result<Vec<(usize, usize)>, git2::Error> {
+    let mut unchanged = Vec::new();
+
+    let patch = Patch::from_blobs(
+        parent_blob,
+        None,
+        blob,
+        None,
+        None,
+    )?;
+
+    if patch.delta().flags().is_binary() {
+        return Ok(unchanged);
+    }
+
+    fn add_delta(lineno: usize, delta: i32) -> usize {
+        ((lineno as i32) + delta) as usize
+    }
+
+    let mut latest_line : usize = 0;
+    let mut delta : i32 = 0;
+
+    for hunk_index in 0..patch.num_hunks() {
+        for line_index in 0..patch.num_lines_in_hunk(hunk_index)? {
+            let line = patch.line_in_hunk(hunk_index, line_index)?;
+
+            if let Some(lineno) = line.new_lineno() {
+                let lineno = lineno as usize;
+                for i in latest_line..lineno - 1 {
+                    unchanged.push((i, add_delta(i, delta)));
+                }
+                latest_line = (lineno - 1) + 1;
+            }
+
+            match line.origin() {
+                '+' => delta -= 1,
+                '-' => delta += 1,
+                ' ' => {
+                    assert_eq!(line.old_lineno().unwrap() as usize, add_delta(line.new_lineno().unwrap() as usize, delta));
+                    unchanged.push(((line.new_lineno().unwrap() - 1) as usize, (line.old_lineno().unwrap() - 1) as usize));
+                }
+                _ => (),
+            };
+        }
+    }
+
+    let linecount = count_lines(blob);
+    for i in latest_line..linecount {
+        unchanged.push((i, add_delta(i, delta)));
+    }
+    Ok(unchanged)
+}
+
+fn blame_for_path(
+    file_movement: &HashMap<Oid, HashMap<Oid, PathBuf>>,
+    git_repo: &git2::Repository,
+    commit: &git2::Commit,
+    blame_repo: &git2::Repository,
+    blame_parents: &[git2::Commit],
+    path: &Path
+) -> Result<String, git2::Error> {
+    let blob = commit.tree()?.get_path(path)?.to_object(git_repo)?.peel_to_blob()?;
+    let linecount = count_lines(&blob);
+    // TODO: drop this author field entirely, I don't think it gets consumed by anything
+    let commit_id = commit.id();
+    let author = commit.author().name().map(|n| n.nfkd().filter(|c| c.is_ascii()).collect::<String>()).unwrap_or_else(String::new);
+    let mut blame = Vec::new();
+    for line in 1..=linecount {
+        blame.push(format!("{}:%:{}:{}", commit_id, line, author));
+    }
+
+    for (i, parent) in commit.parents().enumerate().rev() {
+        let parent_path = file_movement.get(&parent.id())
+            .and_then(|m| m.get(&blob.id()))
+            .map(|p| p.borrow())
+            .unwrap_or(path);
+        let parent_blob = match parent.tree()?.get_path(parent_path) {
```

Comment:
nit: This may be a bit cleaner early-continuing, but your call:

```rust
let parent = match parent.tree()?.get_path(parent_path) {
    Ok(p) => p,
    Err(..) => continue,
};

if !parent.kind().map_or(false, |t| t == ObjectType::Blob) {
    continue;
}

let parent = parent.to_object(git_repo)?.peel_to_blob()?;
```

Or such.